sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchedulingPlugin.scala

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming.state
+
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.rdd.RDD
+import org.apache.spark.scheduler.{Task, TaskLocation, TaskSchedulingPlugin}
+
+class StateSchedulingPlugin(
```

Comment:
Basically this class schedules stateful tasks like original `getPreferredLocations`. It keeps to assign tasks to the location in previous batch. For first batch, it will evenly assign tasks to all executors.

Reply:
With scheduling plugin, we can have more customized logic, e.g., deal with executor lost, evenly distributing state stores across executors, etc.